### DISCUSSION TITLE
Python 3.13?

### DISCUSSION THREAD

**Discussion by aaron-siegel** [mtkennerly/poetry-dynamic-versioning#198] [General]
@mtkennerly , can you comment on the timeline for Python 3.13 support? Since poetry-dynamic-versioning is on the critical path for dependency management for projects using it, we're effectively blocked on supporting 3.13 (unless there's an obvious workaround I've overlooked). Thanks!

**Comment by mtkennerly** [comment]
Hi! I'll try adding 3.12 and 3.13 to the test matrix and see what errors it gets (if any). I wasn't expecting any problems in this project with using 3.13, but I'll confirm.

For reference, is there a particular error you've run into?

**Reply by aaron-siegel** [reply]
Running on a clean conda environment, Python 3.13.0, Mac OS 14.4.1 (M3 Pro ARM):

```
pip install -U pip
pip install poetry==1.8.4  # latest
poetry self add "poetry-dynamic-versioning[plugin]"
```

fails with the following output:

```
Package operations: 4 installs, 20 updates, 0 removals

  - Downgrading certifi (2024.8.30 -> 2024.2.2)
  - Downgrading charset-normalizer (3.4.0 -> 3.3.2)
  - Downgrading idna (3.10 -> 3.7)
  - Downgrading more-itertools (10.5.0 -> 10.2.0)
  - Downgrading urllib3 (2.2.3 -> 2.2.1)
  - Downgrading cffi (1.17.1 -> 1.16.0): Failed

  ChefBuildError

  Backend subprocess exited when trying to invoke build_wheel
  
  running bdist_wheel
  running build
  running build_py
  creating build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/_imp_emulation.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/backend_ctypes.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/error.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/setuptools_ext.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/_shimmed_dist_utils.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/__init__.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/cffi_opcode.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/vengine_gen.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/pkgconfig.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/model.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/ffiplatform.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/api.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/vengine_cpy.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/commontypes.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/lock.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/recompiler.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/cparser.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/verifier.py -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/_cffi_include.h -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/parse_c_type.h -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/_embedding.h -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  copying src/cffi/_cffi_errors.h -> build/lib.macosx-11.1-arm64-cpython-313/cffi
  running build_ext
  building '_cffi_backend' extension
  creating build/temp.macosx-11.1-arm64-cpython-313/src/c
  clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -O2 -Wall -fPIC -O2 -isystem /opt/miniconda3/envs/test/include -arch arm64 -fPIC -O2 -isystem /opt/miniconda3/envs/test/include -arch arm64 -DFFI_BUILDING=1 -DUSE__THREAD -DHAVE_SYNC_SYNCHRONIZE -I/usr/include/ffi -I/usr/include/libffi -I/var/folders/hb/qd0dztsj43j_mdb6hbl1gzyc0000gn/T/tmpyk5efame/.venv/include -I/opt/miniconda3/envs/test/include/python3.13 -c src/c/_cffi_backend.c -o build/temp.macosx-11.1-arm64-cpython-313/src/c/_cffi_backend.o -iwithsysroot/usr/include/ffi
  src/c/_cffi_backend.c:4532:22: warning: 'Py_FileSystemDefaultEncoding' is deprecated [-Wdeprecated-declarations]
                       Py_FileSystemDefaultEncoding, &filename_or_null, &flags))
                       ^
  /opt/miniconda3/envs/test/include/python3.13/fileobject.h:22:1: note: 'Py_FileSystemDefaultEncoding' has been explicitly marked deprecated here
  Py_DEPRECATED(3.12) PyAPI_DATA(const char *) Py_FileSystemDefaultEncoding;
  ^
  /opt/miniconda3/envs/test/include/python3.13/pyport.h:251:54: note: expanded from macro 'Py_DEPRECATED'
  #define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
                                                       ^
  src/c/_cffi_backend.c:6121:9: error: call to undeclared function '_PyErr_WriteUnraisableMsg'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
          _PyErr_WriteUnraisableMsg(PyText_AS_UTF8(s), NULL);
          ^
  src/c/_cffi_backend.c:6121:9: note: did you mean 'PyErr_WriteUnraisable'?
  /opt/miniconda3/envs/test/include/python3.13/pyerrors.h:233:18: note: 'PyErr_WriteUnraisable' declared here
  PyAPI_FUNC(void) PyErr_WriteUnraisable(PyObject *);
                   ^
  1 warning and 1 error generated.
  error: command '/usr/bin/clang' failed with exit code 1
  

  at /opt/miniconda3/envs/test/lib/python3.13/site-packages/poetry/installation/chef.py:164 in _prepare
      160│ 
      161│                 error = ChefBuildError("\n\n".join(message_parts))
      162│ 
      163│             if error is not None:
    → 164│                 raise error from None
      165│ 
      166│             return path
      167│ 
      168│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:

Note: This error originates from the build backend, and is likely not a problem with poetry but with cffi (1.16.0) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "cffi (==1.16.0)"'.

  - Downgrading distlib (0.3.9 -> 0.3.8)
  - Downgrading filelock (3.16.1 -> 3.14.0)
  - Downgrading msgpack (1.1.0 -> 1.0.8)
  - Downgrading packaging (24.2 -> 24.0)
  - Downgrading platformdirs (4.3.6 -> 4.2.2)
  - Downgrading pyproject-hooks (1.2.0 -> 1.1.0)
  - Downgrading rapidfuzz (3.10.1 -> 3.9.0)
  - Downgrading requests (2.32.3 -> 2.31.0)
```

**Reply by edgarrmondragon** [reply]
I think the problem is Poetry's dependency resolution forcing a downgrade of `cffi` to 1.16.0 because it's the latest that supports Python 3.7, but it happens to fail to build on Python 3.13. It might be solved by `poetry-dynamic-versioning` dropping support for Python 3.7, but there also could be a better solution.

**Reply by mtkennerly** [reply]
@edgarrmondragon Yeah, that sounds right. Interestingly, Python 3.13 works with `pipx inject`:

```
$ /c/opt/python/3.13/python.exe -m pipx install --suffix='-py3.13' poetry
$ /c/opt/python/3.13/python.exe -m pipx inject poetry-py3.13 "poetry-dynamic-versioning[plugin]"
$ poetry-py3.13 build
Building dependency-dynamic (1.4.1.post7.dev0+d775f6b)
```

But `poetry-py3.13 self add "poetry-dynamic-versioning[plugin]"` doesn't work because it resolves a few downgrades that fail. For me, on Windows 11, it fails on these:

* `Downgrading msgpack (1.1.0 -> 1.0.7)`
* `Downgrading rapidfuzz (3.10.1 -> 3.5.2)`

...with:

```
  Traceback (most recent call last):
    File "C:\Users\mtken\.local\pipx\venvs\poetry-py3-13\Lib\site-packages\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
      main()
      ~~~~^^
    File "C:\Users\mtken\.local\pipx\venvs\poetry-py3-13\Lib\site-packages\pyproject_hooks\_in_process\_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
                               ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\mtken\.local\pipx\venvs\poetry-py3-13\Lib\site-packages\pyproject_hooks\_in_process\_in_process.py", line 251, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
             ~~~~~~~~~~~~~~^^
    File "C:\Users\mtken\.local\pipx\venvs\poetry-py3-13\Lib\site-packages\pyproject_hooks\_in_process\_in_process.py", line 74, in _build_backend
      ep = os.environ['PEP517_BUILD_BACKEND']
           ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
    File "<frozen os>", line 716, in __getitem__
  KeyError: 'PEP517_BUILD_BACKEND'
```

I guess dropping Python 3.7 is fine since Poetry already dropped it. Poetry 1.8.4 is `python = "^3.8"`, and it looks like the next release will be `python = "^3.9"`: https://github.com/python-poetry/poetry/pull/9692 . I'll continue working on this tomorrow and try to release a fix soon.

**Reply by mtkennerly** [reply]
tl;dr: If you're changing Python versions, you should delete `%APPDATA%/pypoetry/poetry.lock` (or the equivalent on your OS).

I found that, even if I removed all other dependencies from pyproject.toml and only bumped the Python version:

```toml
[tool.poetry.dependencies]
python = "^3.8"
```

...Poetry still wanted to downgrade packages (and the `msgpack` downgrade would fail):

```
$ /c/opt/python/3.13/python.exe -m pipx install --suffix='-py3.13' poetry
installed package poetry 1.8.4 (poetry-py3.13), installed using Python 3.13.0

$ poetry-py3.13 build
  - Built poetry_dynamic_versioning-1.4.1-py3-none-any.whl

$ poetry-py3.13 self add /c/git/poetry-dynamic-versioning/dist/poetry_dynamic_versioning-1.4.1-py3-none-any.whl
Updating dependencies
Resolving dependencies... (0.2s)

Package operations: 1 install, 20 updates, 0 removals

  - Downgrading certifi (2024.8.30 -> 2023.11.17)
  - Downgrading charset-normalizer (3.4.0 -> 3.3.2)
  - Downgrading idna (3.10 -> 3.4)
  - Downgrading more-itertools (10.5.0 -> 10.1.0)
  - Downgrading urllib3 (2.2.3 -> 1.26.18)
  - Downgrading distlib (0.3.9 -> 0.3.7)
  - Downgrading filelock (3.16.1 -> 3.13.1)
  - Downgrading msgpack (1.1.0 -> 1.0.7): Failed
  - Downgrading packaging (24.2 -> 23.2)
  - Downgrading platformdirs (4.3.6 -> 3.11.0)
  - Downgrading jaraco-classes (3.4.0 -> 3.3.0)
  - Downgrading pyproject-hooks (1.2.0 -> 1.0.0)
  - Downgrading pywin32-ctypes (0.2.3 -> 0.2.2)
  - Downgrading rapidfuzz (3.10.1 -> 3.5.2)
  - Downgrading requests (2.32.3 -> 2.31.0)
```

Then I remembered that `%APPDATA%/pypoetry/poetry.lock` exists and was probably causing Poetry to get confused. Once I deleted it, everything started working:

```
$ /c/opt/python/3.13/python.exe -m pipx install --suffix='-py3.13' poetry

$ poetry-py3.13 self add "poetry-dynamic-versioning[plugin]"
Using version ^1.4.1 for poetry-dynamic-versioning

Updating dependencies
Resolving dependencies... (0.9s)

Package operations: 4 installs, 0 updates, 0 removals

  - Installing markupsafe (3.0.2)
  - Installing dunamai (1.22.0)
  - Installing jinja2 (3.1.4)
  - Installing poetry-dynamic-versioning (1.4.1)

Writing lock file

$ poetry-py3.13 build
Building dependency-dynamic (1.4.1.post7.dev0+d775f6b)
```

Please try this and let me know if it works for you.

**Reply by aaron-siegel** [reply]
Indeed, that seems to have worked! Thank you! I did not even know there was an `%APPDATA%/pypoetry/poetry.lock` separate from our project `poetry.lock`. Thanks for your help.

(In case anyone else finds this thread: on Mac OS the offending file is `~/Library/Application Support/pypoetry/poetry.lock`.)

**Reply by mtkennerly** [reply]
Great! I've opened a ticket with Poetry to ask about this as well: https://github.com/python-poetry/poetry/issues/9841
